blackbird_python/blackbird/tests/test_listener.py

```diff
@@ -346,6 +347,313 @@ def test_target_arg(self, parse_input):
 
 
+class TestParsingInclude:
+    """Tests for the `include` statement"""
+
+    @pytest.mark.skipif(sys.version_info < (3, 6), reason="tmpdir fixture requires Python >=3.6")
+    def test_include_program(self, parse_input, tmpdir):
+        """Test including a non-templated program"""
+        program = textwrap.dedent(
+            """
+            name CustomOperation
+            version 0.0
+            float alpha = 0.3423
+            Coherent(alpha, sqrt(pi)) | 0
+            MeasureFock() | 0
+            """
+        )
+
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            f.write(program)
+
+        test_include = textwrap.dedent(
+            """
+            name test_include
+            version 0.0
+            include "{}"
+            CustomOperation | 1
+            """
+        ).format(filename)
+
+        bb = parse_input(test_include, cwd=tmpdir)
+
+        expected = [
+            {"op": "Coherent", "args": [0.3423, np.sqrt(np.pi)], "kwargs": {}, "modes": [1]},
+            {"op": "MeasureFock", "args": [], "kwargs": {}, "modes": [1]},
+        ]
+
+        assert bb.operations == expected
+
+    @pytest.mark.skipif(sys.version_info < (3, 6), reason="tmpdir fixture requires Python >=3.6")
+    def test_include_template(self, parse_input, tmpdir):
+        """Test including a templated program"""
+        template = textwrap.dedent(
+            """
+            name CustomOperation
+            version 0.0
+            BSgate({theta}, pi/2) | [0, 1]
+            Rgate({phi}) | 0
+            """
+        )
+
+        filename = tmpdir.join("test.xbb")
+
+        with open(filename, "w") as f:
+            f.write(template)
+
+        test_include = textwrap.dedent(
+            """
+            name test_include
+            version 0.0
+            include "{}"
+            CustomOperation(theta=0.54, phi=0.1) | [2, 1]
+            """
+        ).format(filename)
+
+        bb = parse_input(test_include, cwd=tmpdir)
+
+        expected = [
+            {"op": "BSgate", "args": [0.54, np.pi / 2], "kwargs": {}, "modes": [2, 1]},
+            {"op": "Rgate", "args": [0.1], "kwargs": {}, "modes": [2]},
+        ]
+
+        assert bb.operations == expected
+
+    @pytest.mark.skipif(sys.version_info < (3, 6), reason="tmpdir fixture requires Python >=3.6")
+    def test_nested_include(self, parse_input, tmpdir):
+        """Test nested includes"""
+        template1 = textwrap.dedent(
+            """
+            name MachZehnder
+            version 0.0
+            BSgate({theta}, pi/2) | [0, 1]
+            Rgate({phi}) | 0
+            """
+        )
+
+        filename1 = tmpdir.join("mach_zehnder.xbb")
+
+        with open(filename1, "w") as f:
+            f.write(template1)
+
+        template2 = textwrap.dedent(
+            """
+            name Chip15
+            version 0.0
+            include "{}"
+            S2gate({{sq}}) | [0, 1]
```

Comment:
What does double curly brackets mean?

Reply:
It's because Python by default treats `{}` as format strings 🙁 So `{{}}` is how you escape curly brackets in format strings, and it will just render as `{sq}`.